File: legacy/src/main/java/com/redbox/domain/auth/util/JWTUtil.java
package com.redbox.domain.auth.util;

import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import io.jsonwebtoken.Claims;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import javax.crypto.SecretKey;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.util.Date;

@Component
public class JWTUtil {

    private final SecretKey secretKey;

    // 암호화 키 불러오기
    public JWTUtil(@Value("${jwt_secret}") String secret) {
        this.secretKey = new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), SignatureAlgorithm.HS256.getJcaName());
    }

    // userId 추출 메소드 추가
    public Long getUserId(String token) {
        Claims claims = parseToken(token);
        return claims.get("userId", Long.class);
    }

    // email 추출
    public String getEmail(String token) {
        Claims claims = parseToken(token);
        return claims.get("email", String.class);
    }

    // role 추출
    public String getRole(String token) {
        Claims claims = parseToken(token);
        return claims.get("role", String.class);
    }

    // category 추출 (access / refresh 구분)
    public String getCategory(String token) {
        Claims claims = parseToken(token);
        return claims.get("category", String.class);
    }

    // 토큰 만료 여부 확인
    public Boolean isExpired(String token) {
        Claims claims = parseToken(token);
        return claims.getExpiration().before(new Date());
    }

    // JWT 생성
    public String createJwt(String category, Long userId, String email, String role, Long expiredMs) {
        return Jwts.builder()
                .claim("category", category)
                .claim("userId", userId)    // userId 추가
                .claim("email", email)
                .claim("role", role)
                .setIssuedAt(new Date(System.currentTimeMillis())) // 발급 시간
                .setExpiration(new Date(System.currentTimeMillis() + expiredMs)) // 만료 시간
                .signWith(secretKey, SignatureAlgorithm.HS256) // 서명
                .compact();
    }

    // 토큰 파싱
    private Claims parseToken(String token) {
        return Jwts.parserBuilder()
                .setSigningKey(secretKey)
                .build()
                .parseClaimsJws(token)
                .getBody();
    }
}

File: migration/src/main/kotlin/com/redbox/global/auth/util/JWTUtil.kt
package com.redbox.global.auth.util

import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.nio.charset.StandardCharsets
import java.util.*
import javax.crypto.SecretKey
import javax.crypto.spec.SecretKeySpec

@Component
class JWTUtil(
    @Value("\${jwt_secret}") secret: String
) {
    private val secretKey: SecretKey =
        SecretKeySpec(
            secret.toByteArray(StandardCharsets.UTF_8),
            SignatureAlgorithm.HS256.jcaName
        )

    // userId 추출 메소드
    fun getUserId(token: String): Long {
        val claims = parseToken(token)
        val userId = claims["userId"]

        return when (userId) {
            is Int -> {
                userId.toLong() // Integer로 저장된 경우 Long으로 변환
            }
            else -> {
                userId as Long // 원래 Long이면 그대로 반환
            }
        }
    }


    // email 추출
    fun getEmail(token: String): String {
        val claims = parseToken(token)
        return claims["email", String::class.java]
    }

    // role 추출
    fun getRole(token: String): String {
        val claims = parseToken(token)
        return claims["role", String::class.java]
    }

    // category 추출 (access / refresh 구분)
    fun getCategory(token: String): String {
        val claims = parseToken(token)
        return claims["category", String::class.java]
    }

    // 토큰 만료 여부 확인
    fun isExpired(token: String): Boolean {
        val claims = parseToken(token)
        return claims.expiration.before(Date())
    }

    // JWT 생성
    fun createJwt(
        category: String,
        userId: Long,
        email: String,
        role: String,
        expiredMs: Long
    ): String {
        return Jwts.builder().apply {
            claim("category", category)
            claim("userId", userId)
            claim("email", email)
            claim("role", role)
            setIssuedAt(Date(System.currentTimeMillis()))
            setExpiration(Date(System.currentTimeMillis() + expiredMs))
            signWith(secretKey, SignatureAlgorithm.HS256)
        }.compact()
    }

    // 토큰 파싱
    private fun parseToken(token: String): Claims {
        return Jwts.parserBuilder()
            .setSigningKey(secretKey)
            .build()
            .parseClaimsJws(token)
            .body
    }
}
